### ISSUE TITLE
Apple // joystick driver does not delay between reading the two axes

### ISSUE THREAD

**Issue by jeremysrand** [cc65/cc65#1043]
Because of the way that the joystick works on the Apple //, there must be a 3ms delay after reading one axis before reading the other as documented in this Apple // tech note:

http://www.gno.org/pub/apple2/doc/apple/technotes/aiie/tn.aiie.006

The current implementation of the joystick driver on cc65 does not do this.  It reads the X axis and then immediately reads the Y axis.  On most (all?) emulators, this is not generally a problem and I can confirm that Virtual ][ does not care about these reads done back-to-back.  But on real HW, the read of the Y axis may be wrong.  If you put the joystick in the extreme lower right corner, you should get a result as though it is in the upper right corner because the Y-axis is read incorrectly.

To demonstrate this, I coded this in C:

    extern char myReadJoy(unsigned char joystick);

    void pollJoystick(void)
    {
        char joyDriverMask;
        char newMask;
        char ch;
    
        do {
            joyDriverMask = joy_read(0);
            printf("Driver - %c%c",
                   (JOY_UP(joyDriverMask) ? 'U' : (JOY_DOWN(joyDriverMask) ? 'D' : 'C')),
                   (JOY_LEFT(joyDriverMask) ? 'L' : (JOY_RIGHT(joyDriverMask) ? 'R' : 'C')));
            
            newMask = myReadJoy(0);
            printf("    New - %c%c\n",
                   (JOY_UP(newMask) ? 'U' : (JOY_DOWN(newMask) ? 'D' : 'C')),
                   (JOY_LEFT(newMask) ? 'L' : (JOY_RIGHT(newMask) ? 'R' : 'C')));
        
            if (kbhit())
            {
                ch = cgetc();
                if ((ch == 'q') ||
                    (ch == 'Q'))
                    return;
            }
        } while (1);
    }

This will display "U", "D" or "C" for the vertical axis to say Y is up, down or centre and "L", "R" or "C" for the horizontal axis to say X is left, right or centre respectively.  Along with a binary which reproduces this behaviour, I am also attaching a potential fix for this in assembly.  It is called in the C snippet above as myReadJoy().

[a2joystick.zip](https://github.com/cc65/cc65/files/4720947/a2joystick.zip)

What the proposed fix does is only read one of X or Y on a poll.  The other axis will always return the position it was in on the previous poll.  I think this is the best solution to address the HW limitation.  The only other solution given the cc65 joystick API I think is to insert a 3ms delay or at least a sufficient delay to allow the capacitor to discharge.

Based on feedback, I would be willing to submit a pull request similar to my proposed solution shown in the joyread.s file in the attached zip file.  There is probably a better way to accomplish this and feedback is definitely welcome.

If I can do anything to clarify the issue or help you reproduce it, let me know.

**Comment by oliverschmidt** [comment]
Hi Jeremy,

> The current implementation of the joystick driver on cc65 does not do this. It reads the X axis and then immediately reads the Y axis.

Thanks for the detailed problem description. To be honest I was wondering why all those years nobody complained about it. That broken driver was already there when I first looked at cc65 - which can be measured by decades by now ;-)

> There is probably a better way to accomplish this and feedback is definitely welcome.

From my POV the only viable way to solve the issue is to not use the PREAD ROM routine at all. Rather a custom routine (based on the knowledge in the very technote you linked) needs to be implemented which reads both paddle circuits at the same time. This was/is the usual way to handle the issue. And that's the reason why every Apple II accelerator slows down on access to $C070. Not doing so would break too many existing programs.

The only issue with reading both paddles at the same time is that the reading loop takes longer which means that the resolution of the paddle position is reduced (from the 0..255 of PREAD). However, when the paddle position is only read to emulate a digital joystick then this is obviously not relevant (as even 0..2 for 'left'/'up' vs 'neutral' vs 'right'/'down' would be enough).

The naive approach to reading both paddles would be to start reading both and as soon as one paddle is triggered continue to read the other until that is triggered too. Additionally the loop needs to stop for sure after a certain amount of time as it otherwise hangs "for ever" of no paddle is connected at all. So this seems rather complicated - especially given that the loop in question should consume the same amount of cycles no matter if one paddle was already triggered.

However, there's a much simpler way to go about it: Just always run the loop for a fixed amount of time and simply record for both paddles when they were triggered. This approach has several benefits:

* Reading the paddles always takes the same amount of time, no matter how the paddles are positioned. This may be relevant for games doing animation.

* There's no special handling necessary for the no-paddle-connected scenario.

* When emulating a digital joystick the loop can end as soon the result maps to 'right' or 'down' as it doesn't matter at all if the paddle is positioned "even further" 'right' or 'down' than the threshold value. This makes the overall amount of time necessary to read the paddles smaller.

So as you see I have a pretty detailed understanding of how the driver should work from my POV. However, I never had the motivation to actually implement it that way - and test it on real hardware (ideally with several types of joysticks).

You seem to be interested enough in the topic to create a test program and a fix proposal. So if you are interested in collaborating with my on a "correct" driver please let me know. My proposal would be that I implement the general algorithm described above and that you test/debug it and provide the best values for the 'left'/'up' threshold and the loop end condition (aka 'right'/'down' threshold).

Regards,
Oliver

**Comment by jeremysrand** [comment]
I can definitely help by testing and tune the thresholds.  I have a //e and a //gs both with CH joysticks.  We can use that HW to tune the thresholds I think.  I will also plan to test without a joystick connected to make sure we don’t hang in the loop forever in that case.

Once we have an implementation that we think is ready with thresholds that we like, I can also share a disk image on the Apple // slack group and try to convince a few other people to run a test program to make sure it works on a greater variety of machines and joysticks.  Does that sound like a reasonable plan?  

> On Jun 3, 2020, at 10:44 AM, Oliver Schmidt <notifications@github.com> wrote:
> 
> 
> Hi Jeremy,
> 
> The current implementation of the joystick driver on cc65 does not do this. It reads the X axis and then immediately reads the Y axis.
> 
> Thanks for the detailed problem description. To be honest I was wondering why all those years nobody complained about it. That broken driver was already there when I first looked at cc65 - which can be measured by decades by now ;-)
> 
> There is probably a better way to accomplish this and feedback is definitely welcome.
> 
> From my POV the only viable way to solve the issue is to not use the PREAD ROM routine at all. Rather a custom routine (based on the knowledge in the very technote you linked) needs to be implemented which reads both paddle circuits at the same time. This was/is the usual way to handle the issue. And that's the reason why every Apple II accelerator slows down on access to $C070. Not doing so would break too many existing programs.
> 
> The only issue with reading both paddles at the same time is that the reading loop takes longer which means that the resolution of the paddle position is reduced (from the 0..255 of PREAD). However, when the paddle position is only read to emulate a digital joystick then this is obviously not relevant (as even 0..2 for 'left'/'up' vs 'neutral' vs 'right'/'down' would be enough).
> 
> The naive approach to reading both paddles would be to start reading both and as soon as one paddle is triggered continue to read the other until that is triggered too. Additionally the loop needs to stop for sure after a certain amount of time as it otherwise hangs "for ever" of no paddle is connected at all. So this seems rather complicated - especially given that the loop in question should consume the same amount of cycles no matter if one paddle was already triggered.
> 
> However, there's a much simpler way to go about it: Just always run the loop for a fixed amount of time and simply record for both paddles when they were triggered. This approach has several benefits:
> 
> 	• Reading the paddles always takes the same amount of time, no matter how the paddles are positioned. This may be relevant for games doing animation.
> 
> 	• There's no special handling necessary for the no-paddle-connected scenario.
> 
> 	• When emulating a digital joystick the loop can end as soon the result maps to 'right' or 'down' as it doesn't matter at all if the paddle is positioned "even further" 'right' or 'down' than the threshold value. This makes the overall amount of time necessary to read the paddles smaller.
> 
> So as you see I have a pretty detailed understanding of how the driver should work from my POV. However, I never had the motivation to actually implement it that way - and test it on real hardware (ideally with several types of joysticks).
> 
> You seem to be interested enough in the topic to create a test program and a fix proposal. So if you are interested in collaborating with my on a "correct" driver please let me know. My proposal would be that I implement the general algorithm described above and that you test/debug it and provide the best values for the 'left'/'up' threshold and the loop end condition (aka 'right'/'down' threshold).
> 
> Regards,
> Oliver
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 

--
Jeremy Rand
jeremy@rand-family.com

**Comment by oliverschmidt** [comment]
> [...] Does that sound like a reasonable plan?

It does :-) Stay tuned...

**Comment by oliverschmidt** [comment]
Hi Jeremy,

https://github.com/cc65/cc65/commit/20a9c0c336387306fa3bdad9d46a79911de77f4c implements the  approach I described above. I tested it with my
 
![image](https://user-images.githubusercontent.com/2664009/83945155-0c22ca00-a809-11ea-8b10-d23a26984338.png)

type joystick on my IIgs both in 'nomal' and 'fast' speed. I wasn't able to test that

* joy_count() returns 2 on all machines but the //c where it returns 1
* the second joystick has an actually working JOY_BTN_2 on the IIgs

Apart from that https://github.com/cc65/cc65/blob/master/libsrc/apple2/joy/a2.stdjoy.s#L22-L23 are the two thresholds in question. I didn't compute anything but rather just experimented with my joystick until the values seemed fine to me.

As mentioned I presume that accelerators don't require special handling - and to be honest I'd **really** like to stay clear form accelerator specific coding in a cc65 general purpose driver. However, if the //c Plus would require special handling I might (try to) add it as it's an original Apple machine.

As discussed I'd appreciate if you could take checking the validity of the driver from here...

Regards,
Oliver

**Comment by jeremysrand** [comment]
Great!  I will start testing tomorrow evening. 

--
Jeremy Rand

**Comment by jeremysrand** [comment]
I have written a cc65 joystick driver test program for the Apple // and it is available here:

https://github.com/jeremysrand/joytest

Based on my testing, the driver is performing well on my //e and //gs, both with CH joysticks.  I am finding that the X axis goes from left to centre at about 13 (on the 0-255 scale) and from centre to right at about 222.  On the Y axis, I get 14 between up and centre and 223 for down and centre.  These seem like good thresholds.  The position is stable in all relevant cases.  I have a Transwarp GS In the //gs and did see some instability with the accelerator on where it would sometimes not hold a position well and maybe bounce between left and down/left when held still.  But that went away when I changed the speed to the original shipping values of 2.8MHz or 1MHz.  I know that accelerator support is outside the scope but honestly, it worked pretty well at 12.5MHz with just a bit of instability in the position at times.

I would be interested in feedback from you about the test program before I take it to the larger community.  Any suggestions for the readme or the program itself to get better test results?


> On Jun 6, 2020, at 11:37 PM, Jeremy Rand <jeremy@rand-family.com> wrote:
> 
> Great!  I will start testing tomorrow evening. 
> 
> --
> Jeremy Rand
> 
> 
> 
>> On Jun 6, 2020, at 9:39 AM, Oliver Schmidt <notifications@github.com> wrote:
>> 
>> ﻿
>> 
>> Hi Jeremy,
>> 
>> 20a9c0c implements the approach I described above. I tested it with my
>> 
>> 
>> type joystick on my IIgs both in 'nomal' and 'fast' speed. I wasn't able to test that
>> 
>> 	• joy_count() returns 2 on all machines but the //c where it returns 1
>> 	• the second joystick has an actually working JOY_BTN_2 on the IIgs
>> Apart from that https://github.com/cc65/cc65/blob/master/libsrc/apple2/joy/a2.stdjoy.s#L22-L23 are the two thresholds in question. I didn't compute anything but rather just experimented with my joystick until the values seemed fine to me.
>> 
>> As mentioned I presume that accelerators don't require special handling - and to be honest I'd really like to stay clear form accelerator specific coding in a cc65 general purpose driver. However, if the //c Plus would require special handling I might (try to) add it as it's an original Apple machine.
>> 
>> As discussed I'd appreciate if you could take checking the validity of the driver from here...
>> 
>> Regards,
>> Oliver
>> 
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub, or unsubscribe.
>> 

--
Jeremy Rand
jeremy@rand-family.com

**Comment by oliverschmidt** [comment]
> [...] Based on my testing, the driver is performing well on my //e and //gs, [...] These seem like good thresholds.  The position is stable in all relevant cases.

Great :-)

> I would be interested in feedback from you about the test program before I take it to the larger community.

Maybe I'd number the joysticks 1..2 instead of 0..1.

Relatively often the "Highest Up" / "Highest Left" switches to 255. As far as I understand this doesn't make sense / is undesirable. But to be honest I don't understand why this happens.

> Any suggestions for the readme or the program itself to get better test results?

I think you can omit the section "Joystick Disconnected:" from your readme. The way I implemented the driver it conceptually can't "hang".

Thanks for the effort you put into hardening the driver!

**Comment by jeremysrand** [comment]
I just made version 0.3 available from the git repository:

https://github.com/jeremysrand/joytest

I changed the joystick numbers to 1 and 2 and removed the disconnected joystick test from the readme.  I also include the raw position from PREAD and I changed the timing of the polling between the cc65 driver and the two calls to PREAD.

I am not seeing the jumping to 255 with highest up or highest left.  It is possible that the changes to the timing of the polls may make that go away I guess.  The raw PREAD position output should be helpful also.  Perhaps you will see 255 values showing up periodically when pushed up or to the left.  Is the position on the left side from the cc65 driver stable and holding steady on the left/up position when this happens?

If you can try 0.3 and see if it addresses some or all of the problems you reported, that would be great.  Thanks.

> On Jun 11, 2020, at 9:20 AM, Oliver Schmidt <notifications@github.com> wrote:
> 
> 
> [...] Based on my testing, the driver is performing well on my //e and //gs, [...] These seem like good thresholds. The position is stable in all relevant cases.
> 
> Great :-)
> 
> I would be interested in feedback from you about the test program before I take it to the larger community.
> 
> Maybe I'd number the joysticks 1..2 instead of 0..1.
> 
> Relatively often the "Highest Up" / "Highest Left" switches to 255. As far as I understand this doesn't make sense / is undesirable. But to be honest I don't understand why this happens.
> 
> Any suggestions for the readme or the program itself to get better test results?
> 
> I think you can omit the section "Joystick Disconnected:" from your readme. The way I implemented the driver it conceptually can't "hang".
> 
> Thanks for the effort you put into hardening the driver!
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 

--
Jeremy Rand
jeremy@rand-family.com

**Comment by oliverschmidt** [comment]
I think your test program suffers from the very problem the old cc65 driver suffered from. And it is behaving differently depending on the IIgs speed setting. But to be honest I'm not exactly sure what the actual goal for displaying the values is. Anyhow my experience with 0.4:

Normal speed:
* For Y axis Lowest Down stays N/A no matter what I do.

Fast speed:
* Highest Left / Highest Up still easily get 255.
* Y position "wraps around" when moving the stick all the way down to the end and is somewhere between 15..35.

My suggestions would be to have to program either

* only call into the driver and display what it currently does in the upper left corner of the screen

or

* duplicate new current driver code with the difference to allow the user to adjust the two threshold values in question and provide you with feedback if it was necessary to adjust the thresholds and what values worked best for the user

Just my two cents,
Oliver

**Comment by jeremysrand** [comment]
I agree.  I definitely have over-thought this.  I have published version 0.5 on GitHub and it only polls the joystick using the cc65 driver.  I believe this should be a more reliable test program.

> On Jun 14, 2020, at 8:26 AM, Oliver Schmidt <notifications@github.com> wrote:
> 
> 
> I think your test program suffers from the very problem the old cc65 driver suffered from. And it is behaving differently depending on the IIgs speed setting. But to be honest I'm not exactly sure what the actual goal for displaying the values is. Anyhow my experience with 0.4:
> 
> Normal speed:
> 
> 	• For Y axis Lowest Down stays N/A no matter what I do.
> Fast speed:
> 
> 	• Highest Left / Highest Up still easily get 255.
> 	• Y position "wraps around" when moving the stick all the way down to the end and is somewhere between 15..35.
> My suggestions would be to have to program either
> 
> 	• only call into the driver and display what it currently does in the upper left corner of the screen
> or
> 
> 	• duplicate new current driver code with the difference to allow the user to adjust the two threshold values in question and provide you with feedback if it was necessary to adjust the thresholds and what values worked best for the user
> Just my two cents,
> Oliver
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 

--
Jeremy Rand
jeremy@rand-family.com

**Comment by oliverschmidt** [comment]
Version 0.5 works just fine for the (few) scenarios I can test here.

BTW: made the driver a bit smaller with https://github.com/cc65/cc65/commit/e0a0b2dc25bfc55e9f9517f82d7bb4b158c065f3. It is supposed to behave identical, but just to be sure you may want to rebuild your test program against the lastest driver.

**Comment by jeremysrand** [comment]
I just published version 0.6 which uses the latest cc65 and includes the updated driver code.  I plan to distribute this build wider for more testing on real HW tomorrow.  Let me know if you see any issues with 0.6 or if there is any reason to delay getting the community involved in testing now.  Thanks!


> On Jun 15, 2020, at 1:37 PM, Oliver Schmidt <notifications@github.com> wrote:
> 
> 
> Version 0.5 works just fine for the (few) scenarios I can test here.
> 
> BTW: made the driver a bit smaller with e0a0b2d. It is supposed to behave identical, but just to be sure you may want to rebuild your test program against the lastest driver.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 

--
Jeremy Rand
jeremy@rand-family.com

**Comment by oliverschmidt** [comment]
From my POV version 0.6 is ready fro prime time!

**Comment by jeremysrand** [comment]
I have announced the joystick test program on a couple of community forums and hope to hear from people soon with their feedback.  I will let you know how it goes.

> On Jun 17, 2020, at 1:46 PM, Oliver Schmidt <notifications@github.com> wrote:
> 
> 
> From my POV version 0.6 is ready fro prime time!
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 

--
Jeremy Rand
jeremy@rand-family.com

**Comment by oliverschmidt** [comment]
Without any negative feedback so far I consider this fixed with https://github.com/cc65/cc65/commit/20a9c0c336387306fa3bdad9d46a79911de77f4c.

**Comment by jeremysrand** [comment]
Sorry for the lack of response on this.  I haven’t gotten much feedback so far and most of what I have gotten has been on HW more or less like what I have already tested on.  That said, no negative feedback about the driver.

I am going to try the KansasFest community next to see if I can get more feedback.  No matter what, if there are any issues, I will raise them as a new issue.  Thanks for your work on this much improved driver.


> On Jul 15, 2020, at 1:53 AM, Oliver Schmidt <notifications@github.com> wrote:
> 
> 
> Without any negative feedback so far I consider this fixed with 20a9c0c.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 

--
Jeremy Rand
jeremy@rand-family.com

**Comment by oliverschmidt** [comment]
> That said, no negative feedback about the driver.

:-)

> If there are any issues, I will raise them as a new issue.

Ok.

> Thanks for your work on this much improved driver.

You're welcome! Anyhow, fixing that driver was overdue...
